test/upgrade/upgrade.go

```diff
@@ -42,3 +51,251 @@ func RunMainTest(m *testing.M) {
 		return m.Run()
 	}())
 }
+
+// DurableFeature holds the setup and verify phase of a feature. The "setup" phase should set up
+// the feature. The "verify" phase should only verify its function. This function should be
+// idempotent. Calling this function multiple times should still properly verify the feature
+// (e.g. one call after upgrade, one call after downgrade).
+type DurableFeature struct {
+	SetupF   *feature.Feature
+	setupEnv environment.Environment
+	setupCtx context.Context
+	VerifyF  *feature.Feature
+	Global   environment.GlobalEnvironment
+}
+
+func (fe *DurableFeature) Setup(label string) pkgupgrade.Operation {
+	return pkgupgrade.NewOperation(label, func(c pkgupgrade.Context) {
+		c.T.Parallel()
+		ctx, env := fe.Global.Environment(
+			knative.WithKnativeNamespace(system.Namespace()),
+			knative.WithLoggingConfig,
+			knative.WithTracingConfig,
+			k8s.WithEventListener,
+			// Not managed - namespace preserved.
+		)
+		fe.setupEnv = env
+		fe.setupCtx = ctx
+		env.Test(ctx, c.T, fe.SetupF)
+	})
+}
+
+func (fe *DurableFeature) Verify(label string) pkgupgrade.Operation {
+	return pkgupgrade.NewOperation(label, func(c pkgupgrade.Context) {
+		c.T.Parallel()
+		fe.setupEnv.Test(fe.setupCtx, c.T, fe.VerifyF)
+	})
+}
+
+func (fe *DurableFeature) VerifyTeardown(label string) pkgupgrade.Operation {
+	return pkgupgrade.NewOperation(label, func(c pkgupgrade.Context) {
+		c.T.Parallel()
+		fe.setupEnv.Test(fe.setupCtx, c.T, fe.VerifyF)
+		// Ensures teardown of resources/namespace.
+		fe.setupEnv.Finish()
+	})
+}
+
+func (fe *DurableFeature) SetupVerifyTeardown(label string) pkgupgrade.Operation {
```

Comment:
nit: (same as above)
```suggestion
func (fe *DurableFeature) SetupVerifyAndTeardown(label string) pkgupgrade.Operation {
```